### PR TITLE
Make "Registering of All Segments for a BBox"- feature mag aware

### LINKED ISSUE #8044: Register all segments in this bounding box should be mag-aware
## Detailed Description

- The "Register all segments in this bounding box" functionality has an upper limit on the volume in vx the bounding box is allowed to have. The limit should be in voxels _in the current mag_ and the segmentation in the current mag should be used.

## Context

This would allow to register segments for large objects, for example somata or nuclei which is usually done in a much coarser mag, for example mag 32. With the current mag-unaware limit this is not possible at all.

### PR DESCRIPTION
**Pull request by dieknolle3333**
### URL of deployed dev instance (used for testing):
- https://registerallsegmentsmagaware.webknossos.xyz/

### Steps to test:
- Open an annotation, ideally with multiple segmentation layers
- Open the bounding box tab and select "Register all segments for this bounding box"
- The feature should work as before, registering the segments into a new segment group
- Now use it on a very large BBox that exceeds the maximal volume in mag 1-1-1 (the limit is 512 * 512 * 512 Vx³)
- zoom out to coarser mags and try again to register all segments, until it is possible. This proves that the volume limit is taking the current mag into account
- Register the segments for a bounding box in multiple mags, from coarse to fine. In the corresponding segment groups, some segments from the finer mag should not be moved into the group that was created from the coarser mag. This would prove that the right mag is used to register the segments. (Context: If you register the segments for the same bbox in the same mag twice, two segment groups will be created. As segments can only be present in one group, each segment that is registered again will be moved into the latest "register all segments" segment group. So now, in our case, when a segment remains in the older group, this means that it was not registered again, and that means that the coarser mag was actually used)

- In an annotation with multiple segmentation layers, use this feature and make sure that the currently visible segmentation layer is used. One way to check this is to create a new volume annotation layer in an annotation with another segmentation layer, brush to create one segment, and then register the segments in both segmentation layers. For the newly created segmentation layer, only the brushed segment should be registered. In the other segmentation layer, this should be different.

### TODOs:
- [x] use current mag for volume limit
- [x] use segmentation in current mag to registering segments
- [x] use visible segmentation layer rather than first one. 
    - [x] test me!
    - [x] fix problem that segmentation layer name cannot be found sometimes -> I dont think this had anything todo with the new issue

### Issues:
- fixes #8044 

------
(Please delete unneeded items, merge only when none are left open)
- [x] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change
- [ ] Removed dev-only changes like prints and application.conf edits
- [x] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)
- [ ] Needs datastore update after deployment

### CODE REVIEW

**Comment by daniel-wer** [comment]
Oh and thanks a lot for thinking through and documenting how to properly test this, that was very helpful :pray:

**Comment by dieknolle3333** [comment]
>Feel free to tackle that in this PR or to create a followup issue.

To be honest I stumbled across this case aswell, but I did not understand what made this layer special exactly, so thank you for pointing that out. I'll try to fix it and see how much effort it takes, maybe the lowest effort would be an appropriate error message.

File: frontend/javascripts/oxalis/api/api_latest.ts [hunk 1]
```diff
@@ -656,11 +657,28 @@ class TracingApi {
...
+    const state = Store.getState();
     const maximumVolume = options?.maximumVolume ?? Constants.REGISTER_SEGMENTS_BB_MAX_VOLUME_VX;
     const maximumSegmentCount =
       options?.maximumSegmentCount ?? Constants.REGISTER_SEGMENTS_BB_MAX_SEGMENT_COUNT;
     const shape = Utils.computeShapeFromBoundingBox({ min, max });
-    const volume = Math.ceil(shape[0] * shape[1] * shape[2]);
+
+    const segmentationLayerName = api.data.getVisibleSegmentationLayerName();
+    if (segmentationLayerName == null) {
+      Toast.error("The current segmentation layer could not be found.");
```
[the remark this answers is not in the document]
Comment (dieknolle3333) [reply frontend/javascripts/oxalis/api/api_latest.ts:668]: Yes, totally!

[the remark this answers is not in the document]
Comment (dieknolle3333) [reply frontend/javascripts/oxalis/api/api_latest.ts:668]: Regarding the warning toast and converting it to a console.warn, I fear that users are not regularly checking the console. How about a toast and console.warn? The toast doesnt hurt when this is used for scripting, right?

File: frontend/javascripts/oxalis/api/api_latest.ts [hunk 2]
```diff
@@ -672,19 +690,16 @@ class TracingApi {
...
-    const finestResolution = resolutionInfo.getFinestResolution();
-    // By default, getDataForBoundingBox uses the finest existing magnification.
-    // We use that as strides to traverse the data array properly.
-    const [dx, dy, dz] = finestResolution;
-
-    const data = await api.data.getDataForBoundingBox(segmentationLayerName, {
-      min,
-      max,
-    });
+    const magIndex = getActiveMagIndexForLayer(state, segmentationLayerName);
```
[the remark this answers is not in the document]
Comment (dieknolle3333) [reply frontend/javascripts/oxalis/api/api_latest.ts:693]: makes so much sense!

**Comment by dieknolle3333** [comment]
> I'll try to fix it and see how much effort it takes, maybe the lowest effort would be an appropriate error message.
I think the best solution is to register the segments in the root group :)

File: frontend/javascripts/oxalis/api/api_latest.ts [hunk 3]
```diff
@@ -665,39 +664,45 @@ class TracingApi {
...
     }
 
-    const currentMag = getCurrentResolution(state, segmentationLayerName);
+    const resolutionInfo = getResolutionInfo(
+      getLayerByName(state.dataset, segmentationLayerName).resolutions,
+    );
+    const theoreticalMagIndex = getActiveMagIndexForLayer(state, segmentationLayerName);
+    const existingMagIndex = resolutionInfo.getIndexOrClosestHigherIndex(theoreticalMagIndex);
+    if (existingMagIndex == null) {
+      throw new Error("The index of the current mag could be found.");
```
Comment (daniel-wer) [inline frontend/javascripts/oxalis/api/api_latest.ts:678]:
```suggestion
      throw new Error("The index of the current mag could not be found.");
```

**Review by daniel-wer — APPROVED** [review]
Nicely done! :tada: 

I think adding segments to the root group when a layer is not editable (and has no associated volume tracing), is a good solution, but could you please add a toast warning that the segment list won't be persisted across page reloads, then? You could also mention that in order to do so, the respective layer should be made editable by clicking the button right of the layer name.

Afterwards, feel free to merge :ship:

File: frontend/javascripts/oxalis/api/api_latest.ts (hunk 3 above)
Comment (dieknolle3333) [reply frontend/javascripts/oxalis/api/api_latest.ts:678]: Negations are hard, apparently :sweat_smile:

**Pull request merged**